### ISSUE TITLE
[Bug]: Small vertical line before powerline and box drawing characters in Nerd Fonts

### ISSUE THREAD

**Issue by patrick96** [polybar/polybar#2865]
### Checklist

- [X] I have read the appropriate section in the [contributing guidelines](https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md)
- [X] I believe this issue is a problem with polybar itself and not a misconfiguration on my part
- [X] I have searched for other open and closed [issues](https://github.com/polybar/polybar/issues?q=is%3Aissue) that may have already reported this problem
- [X] I have checked the [known issues](https://github.com/polybar/polybar/wiki/Known-Issues) page for this problem.
- [X] I have followed the [debugging guide](https://github.com/polybar/polybar/wiki/Debugging-your-Config) to narrow down the problem to a minimal config.

### Steps to reproduce

polybar -c config.ini main

### Minimal config

```dosini
[bar/main]
font-0 = "Literation Mono Nerd Font:size=22"
modules-left = mod1 mod2
background = #ffffff
foreground = #000000

width = 100

[module/mod1]
type = custom/text
label = " %{+u}█%{+u}█%{+u}█"

[module/mod2]
type = custom/text
label = █
label-foreground = #dddddd
label-background = #000000
```

### Polybar log

```text
notice: Parsing config file: config.ini
notice: Loading module 'mod1' of type 'custom/text'
notice: Loading module 'mod2' of type 'custom/text'
notice: Loaded 2 modules
notice: Loaded font "Literation Mono Nerd Font:size=22" (name=LiterationMono Nerd Font, offset=0, file=/usr/share/fonts/TTF/Literation Mono Nerd Font Complete.ttf)
```

### Expected behavior

All the characters should be flush without a gap in between.

### Actual behavior

There is a 1px red-ish gap before the box-drawing character with black foregound and a blue-ish gap before the box-drawing character with a gray foreground.

### Window Manager and Version

HerbstluftWM 0.9.5

### Linux Distribution

ArchLinux

### Polybar version

```text
polybar 3.6.3-80-g30e1cc25

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -flto=auto -O3 -DNDEBUG -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto=auto -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override  -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override
```

### Additional Context / Screenshots

![x](https://user-images.githubusercontent.com/2452038/202048284-bbd69732-7039-4dc9-9be9-efd5e9f14f5e.png)

I initially thought this was the same issue as #2785, but the fix for that issue does nothing here.

This only happening when there is a tag in between is a strong indication of a similar underlying problem though.

**Comment by jayywolff** [comment]
I experience similar issues nearly every time the nerd fonts library is updated

on the latest nerd font update, the powerline symbols seem fine in every application expect polybar there's some weird screen tearing effects now on every status update

**Comment by Carlisle96** [comment]
It also happened with me when i updated my system. ( Polybar was already on highest version - it could be that in that update nerd fonts were in there. )

**Comment by jayywolff** [comment]
looks like this could be an issue with x11 or something in the graphics stack, im able to reproduce the issue in gvim

**Comment by Asem-Ashraf** [comment]
The problem still exists, but I was able to identify that it was due to anti-aliasing. If antialias=false for the font, the line goes away but the powerline symbols looks awful. I have powerline symbols on the left and right of each module by the format suffix and prefix, each module has its own color and but the vertical line color is the same across all modules, which lead me to suspect that it is solely anti-aliasing problem of the font. 
![image](https://user-images.githubusercontent.com/73125888/215792135-b91bb183-31bd-47b9-9531-8f9ea3500182.png)

**Comment by jayywolff** [comment]
disabling anti-aliasing doesn't fix it for me in polybar or gvim. I forgot to mention it, but I also don't have any issues with the font glyphs in alacritty only polybar / gvim

**Comment by Asem-Ashraf** [comment]
https://github.com/polybar/polybar/wiki/Formatting#offset-o
This is the method that I followed to offest the characters a to hide the vertical line.
If the same glyph is used more than one time, I would suggest declaring it in variable with the offest one time rather than repeating it across the configuration.

**Comment by Hempward** [comment]
Thanks for the link and suggestion. I'm sure I've sailed past it a few times over the years. Now I can put the section to some use.

**Comment by patrick96** [comment]
@Asem-Ashraf You're right this does look like an anti-aliasing issue. The font renderer inserts pixels with different colors to make edges look smoother. In most cases, this is desirable, but for the straight edges where powerline symbols and the like are supposed to connect together, this creates these small vertical lines.

**Small Example**

```dosini
[bar/main]
font-0 = "Literation Mono Nerd Font:antialias=off:size=17;3"
; font-0 = "Literation Mono Nerd Font:antialias=on:size=17;3"
modules-left = mod1
background = #ffffff
foreground = #000000

width = 200
height = 48

override-redirect = true

[module/mod1]
type = custom/text
label = " %{+u}███%{+u}█ %{+u} "
```

With anti-aliasing:

![polybar-antialias](https://user-images.githubusercontent.com/2452038/236952097-da34951a-6721-4d1d-9ce2-ea0cd27a20eb.png)

Without anti-aliasing:

![polybar-no-antialias](https://user-images.githubusercontent.com/2452038/236952099-8bf0e2a6-2fc4-4a89-96b1-47672af33af4.png)

*For the half-circles you may need to zoom in a bit, the left one has a small line in the middle when anti-aliasing is turned on.*

There are two problems here. For one, as seen in the two circles, adding a formatting tag in between, causes a line to appear.
The reason for that is that the two half-circles are rendered separately (because the formatting tag *could* have changed something, so the rendering is split up into separate chunks with different formatting values) and so each rendering may decide to add this vertical line for anti-aliasing.
If there is no tag in the middle (right circle), the font renderer decides, in this case, that this is not necessary in between the half circles.

The second problem is the anti-aliasing itself. We can see this in the rectangle on the left. The rectangle is made out of full size [block elements](https://en.wikipedia.org/wiki/Block_Elements). Some of them have a formatting tag in between, but not all and still all of them have vertical lines due to anti-aliasing.

I don't see a satisfying solution for this. Some partial solutions:

**Turning off anti-aliasing**

This will solve the issue these vertical lines, but most other stuff will look ugly. As seen above, the half circles will not look very smooth, the same goes for anything else that does not exclusively have vertical and horizontal edges, including text.

@jayywolff You have reported that turning off anti-aliasing does not work for you. However, I think in your case the symbols are also slightly too small and don't fill out the full width, also causing a gap. Try playing around with the size for the powerline symbols.

**Turning off anti-aliasing only for specific characters**

This requires multiple font definitions and manually setting the non-antialias font for all desired characters.
It does solve unrelated characters not looking smooth, but things like powerline symbols which still have no anti-aliasing to get rid of these vertical lines still look ugly.

**Negative 1px offset**

As described above by @Asem-Ashraf, one can manually apply a negative 1px offset before or after these symbols (depending on where the vertical lines occur) to have the subsequent text move back 1 pixel and basically overwrite the 1px vertical line.

This should indeed work as long as the discolored vertical line from anti-aliasing is only 1 pixel wide.
It is however very tedious to do and you also lose 1 pixel from the icons which may make them look slightly off (especially for the half circles).

---

I have been looking into how to solve this in the code and have so far been unsuccessful.

I have tried to render something with **pango**, which I hoped would not have this issue, but I was able to replicate the same behavior:

![pango](https://user-images.githubusercontent.com/2452038/236956545-2c9cf88a-a10c-4f00-af1f-12b316801e1f.png)

This was produced with the following markup string:

```
 ████ <span foreground=\"#ff0000\"></span> 
```

As you can see, simply breaking up the two half circles (on the left) with a zero-width element, has the same effect as in polybar and the full size block elements also all exhibit the same anti-aliasing artifact.

**i3bar** has a similar problem:

![i3bar](https://user-images.githubusercontent.com/2452038/236956764-4295fcea-a736-4cc8-8d9d-9534c3baff63.png)

The full size block elements also have these vertical bars. For some reason the half circles overlap, so we can't do any comparisons there.

So does **tint2**:

![tint2](https://user-images.githubusercontent.com/2452038/236958499-6688af97-99e6-4072-9d37-0c72999843b1.png)

tint2 doesn't seem to have any kind of formatting syntax, so we can't insert anything between the half circles like we did in polybar. But the block elements all also have the vertical lines.

---

I have no idea how to solve this or if it even is solvable. Unless anyone has an idea for how to approach this, I will have to put this on the back burner.
